x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/PruneColumns.java

```diff
@@ -148,24 +96,143 @@ public LogicalPlan apply(LogicalPlan plan) {
         return pl;
     }
 
+    private static LogicalPlan pruneColumns(Aggregate aggregate, AttributeSet.Builder used, boolean inlineJoin) {
+        LogicalPlan p = aggregate;
+
+        var remaining = pruneUnusedAndAddReferences(aggregate.aggregates(), used);
+
+        if (remaining != null) {
+            if (remaining.isEmpty()) {
+                if (inlineJoin) {
+                    p = emptyLocalRelation(aggregate);
+                } else if (aggregate.groupings().isEmpty()) {
+                    // We still need to have a plan that produces 1 row per group.
+                    p = new LocalRelation(
+                        aggregate.source(),
+                        List.of(Expressions.attribute(aggregate.aggregates().getFirst())),
+                        LocalSupplier.of(new Block[] { BlockUtils.constantBlock(PlannerUtils.NON_BREAKING_BLOCK_FACTORY, null, 1) })
+                    );
+                } else {
+                    // Aggs cannot produce pages with 0 columns, so retain one grouping.
+                    Attribute attribute = Expressions.attribute(aggregate.groupings().getFirst());
+                    NamedExpression firstAggregate = aggregate.aggregates().getFirst();
+                    remaining = List.of(new Alias(firstAggregate.source(), firstAggregate.name(), attribute, firstAggregate.id()));
+                    p = aggregate.with(aggregate.groupings(), remaining);
+                }
+            } else {
+                if (inlineJoin && aggregate.groupings().containsAll(remaining)) { // not expecting high groups cardinality
+                    // It's an INLINEJOIN and all remaining attributes are groupings, which are already part of the IJ output (from the
+                    // left-hand side).
+                    if (aggregate.child() instanceof StubRelation stub) {
```

Comment:
I _think_ this condition will need an update when we start supporting filters on aggs... if that guess is valid, I think a TODO is in order here. If you will add that TODO, for some (all, most I hope) I added `// TODO Inlinestats:....` to find them quicker later.